### DISCUSSION TITLE
[1주차]_3장_시스템 설계 면접 공략법

### DISCUSSION THREAD

**Discussion by minju412** [11th-SSAFY-19/large-scale-system-design#6] [General]
.

**Comment by minju412** [comment]
<img src=https://github.com/11th-SSAFY-19/large-scale-system-design/assets/59405576/007aeebc-6107-4510-8d0b-17af2a0d6fe1 width="40%" height="40%"/><br>
45페이지 그림 중, 포스팅 저장 서비스, 포스팅 전송 서비스, 알림 서비스 세 개로 나뉘어져 있는데, MSA로 이해하면 되는걸까요..??

**Reply by minju412** [reply]
웹서버에서 모든걸 처리할 수 없으니 알림을 보내 놓고서 생각하는 느낌. MSA는 아닌 것 같다.
구조가 중요하기보다는 웹서버에서 모든걸 처리하지 않는다 -> 카프카!

**Reply by gurwls0122** [reply]
MSA의 개념에 부합한다고 생각할 수도 있을 것 같아요. 기본적으로 각 기능별로 서버를 나눈 형태처럼 보이기 때문에요..!

이후에 말했던 것처럼 책 스터디 마감할 때쯤 생각을 다시 나눠보는 것도 좋을것같아요.

**Comment by hoshogi** [comment]
한 기업 면접에서 단일 서버, 단일 데이터베이스로 되어 있는 서비스가 성장해서 1,000만 명의 회원이 사용한다고 했을 때, 
어떻게 아키텍쳐를 수정해야 할 지에 대한 질문을 받은 적이 있습니다.
이 질문에 대한 답변을 어떻게 하는 게 좋을지와 답변을 위해 면접관님께 추가적인 질문들을 어떻게 해야 할지 대한 전략들에 대해 함께 얘기해 보고 싶습니다.

**Reply by minju412** [reply]
서비스가 성장했다고 했는데, 어느방향으로 성장한건지 구체적으로 물어봐도 좋을 듯! 이상형 월드컵 기능으로만 천만명이 된거면 부하 분산을 하는게 좋고, 다른 기능이 더 추가되어 성장한거면 MSA.
이상형 월드컵에 조회 기능이 더 많다면 디비 다중화를 고려해도 좋을 듯. + 캐시
-재호-

**Reply by gurwls0122** [reply]
3가지 관점에서 생각할 수 있을 것 같아요!
1. 어플리케이션 관점 : 로드밸런싱(트래픽을 분산 시켜 서버 부하를 줄임) / 어플리케이션 서버 수평 적 확장 / 캐싱(Redis, Memcached와 같은 인메모리 캐시 사용)
2. 데이터베이스 레벨 수정 : 샤딩(데이터베이스 인스턴스 분산) / 레플리카(마스터 - 슬레이브로 읽기 요청 부산) / DB 클러스터링(데이터베이스를 클러스터로 구성)
3. 아키텍처 레벨 수정 : MSA(여러 독립적인 마이크로 서비스로 분리) / 서버리스 아키텍처(AWS Lambda와 같은 서버리스 플랫폼 활용)

하지만, 재호형님 말씀대로 실제 어떤 부분에서 확장해야 될지 명시적으로 물어보는 방식을 사용하는게 좋을것 같아요..!
